### PR TITLE
bootstrap handles BundleIDs, not full bundles

### PR DESCRIPTION
**Pull request by warner**
The core-bootstrap `installBootContracts()` step is responsible for
creating Zoe installations for each "boot contract": contracts that
are included in the initial chain configuration, and instantiated
during bootstrap (currently just `centralSupply` and
`mintHolder`). These contract bundles are specified by config.bundle
entries, which provide a `sourceSpec` that points to the bundle file
on disk.

Previously, this step obtained a full bundle for each contract, in
RAM, and then used `zoe.install(bundle)` to create the Zoe
Installation. When called this way, Zoe must store the full
bundle (about 1MB) in its durable storage, and the bundle appears as a
message argument in lots of places on its way to being evaluated
inside a ZCF vat, which slows things down and consumes more disk
space.

The preferred way to interact with bundles is through BundleCaps. This
commit changes installBootContracts to look up the BundleID of each
boot contract, and submit the ID to `zoe.installBundleID()`. This
allows Zoe to retrieve a (small) BundleCap, and store the bcap in the
installation record. With this approach, the full bundle only appears
once, when the ZCF vat retrieves its contents just prior to
evaluation.

Changing installBootContracts is straightforward, however we have a
number of mock Zoe and mock VatAdminService implementations that
appear during unit tests, and these required changes to accomodate the
new approach. `zoe/tools/fakeVatAdmin.js` provides the primary mock
service, but it is only capable of creating ZCF vats. It was augmented
to understand the full suite of `getBundleIDByName`/etc methods, with
helper tools to populate the simulated tables. Then
`vats/tools/boot-test-utils.js` acquired a wrapper around zoe's mock
service, adding the ability to create non-ZCF vats, and populating the
tables with all the bundles known to the unit tests.

refs #6826
refs #4374

### CODE REVIEW

**Comment by warner** [comment]
cc @mhofman let's see if this helps with the loadgen slogfile size

File: packages/ERTP/src/types-ambient.js
Comment (dckc) [inline packages/ERTP/src/types-ambient.js:282]: FYI, there's a bunch of stuff in `packages/vats/src/core/types.js` that probably belongs in swingset too... ah... we have #4655 for that.

File: packages/vats/src/core/basic-behaviors.js
Comment (dckc) [inline packages/vats/src/core/basic-behaviors.js:300]: yay for less power... or... less direct device access, anyway; I suppose `vatAdminSvc` closes over these 2.

File: packages/vats/src/core/basic-behaviors.js
Comment (dckc) [inline packages/vats/src/core/basic-behaviors.js:316]: :+1:

This seems to be the gist of it...

File: packages/vats/test/devices.js
Comment (dckc) [inline packages/vats/test/devices.js]: There are no functional changes in `getNamedBundleCap`, right? just style?

File: packages/vats/tools/boot-test-utils.js
Comment (dckc) [inline packages/vats/tools/boot-test-utils.js:92]: sneaky!

File: packages/vats/tools/boot-test-utils.js
Comment (dckc) [inline packages/vats/tools/boot-test-utils.js]: consider: `for (const [name, bundle] of Object.entries(bundles)) {`

File: packages/vats/tools/boot-test-utils.js
Comment (dckc) [inline packages/vats/tools/boot-test-utils.js:120]: TIL.

File: packages/zoe/tools/fakeVatAdmin.js
Comment (dckc) [inline packages/zoe/tools/fakeVatAdmin.js:73]: I wonder who else around here could have figured this out. I suppose given a few days some of us might have puzzled it out.

File: packages/SwingSet/src/types-external.js
Comment (dckc) [inline packages/SwingSet/src/types-external.js:374]: I struggle to read the `@callback` syntax.

But I guess you just moved it around; you didn't introduce it here.

**Review by dckc — APPROVED** [review]
great stuff!

`tsc` shows some type conflict with the smart wallet package around `BootstrapSpace`; let me know if you want another set of eyes on that.

File: packages/ERTP/src/types-ambient.js
Comment (warner) [reply packages/ERTP/src/types-ambient.js:282]: yeah, I stopped doing cleanup once `tsc` stopped complaining (at least in the packages that I thought it would notice, apparently there's still more to fix)

File: packages/vats/src/core/basic-behaviors.js
Comment (warner) [reply packages/vats/src/core/basic-behaviors.js:316]: yep, for sure

File: packages/vats/test/devices.js
Comment (warner) [reply packages/vats/test/devices.js]: Oh, sorry, that was a mistake, I temporarily had a console.log, and had to rearrange the function to accomodate it. I'll revert the unnecessary change.

File: packages/vats/tools/boot-test-utils.js
Comment (warner) [reply packages/vats/tools/boot-test-utils.js]: nice, will do

File: packages/SwingSet/src/types-external.js
Comment (gibson042) [inline packages/SwingSet/src/types-external.js]: Is `vatParameters` not required to be an object when supplied?
```suggestion
 * @typedef { { vatParameters?: object, upgradeMessage: string } } VatUpgradeOptions
```
(_and as an aside, I'm growing to hate single-line typedefs because the name is at the end rather than right after the keyword, forcing me to scan lines like `@typedef …` → `@typedef … TypeName` ← `@typedef {typeDescription} TypeName` if I want to understand them—which is especially bad on a phone screen where the single source line spans multiple lines when rendered. I wonder if we should adopt a convention to use `@property` like following typedefs more often, if not "always" [my preference] then maybe "when there is more than one property"?_)

File: packages/SwingSet/src/types-external.js
Comment (gibson042) [inline packages/SwingSet/src/types-external.js]: Is it appropriate to mention "contract" here? I thought vatAdmin supports manipulation of _arbitrary_ vats.
```suggestion
 * @typedef {object} VatAdminFacet
 * A powerful object corresponding with a vat
 * that can be used to upgrade it with new code or parameters,
 * terminate it, or be notified when it terminates.
```

File: packages/SwingSet/src/types-external.js
Comment (gibson042) [inline packages/SwingSet/src/types-external.js]:
```suggestion
 * @property {() => Promise<any>} done
 * returns a promise that will be fulfilled or rejected when the vat is
 * terminated. If the vat terminates with a failure, the promise will be
 * rejected with the reason. If the vat terminates successfully, the
 * promise will fulfill to the completion value.
```

File: packages/SwingSet/src/types-external.js
Comment (gibson042) [inline packages/SwingSet/src/types-external.js:385]: I find myself confusing this type as corresponding with the vatAdmin _service_... can we find a better name for it? Maybe "VatHandle" as in https://en.wikipedia.org/wiki/Handle_(computing) ? I'm not completely satisfied with that, but it does at least communicate the single-vat orientation and suggest some kind of interaction capability.

File: packages/SwingSet/src/types-external.js
Comment (gibson042) [inline packages/SwingSet/src/types-external.js]:
```suggestion
 * @property {ShutdownWithFailure} terminateWithFailure
 * Terminate the vat with a failure reason.
```

File: packages/SwingSet/src/types-external.js
Comment (gibson042) [inline packages/SwingSet/src/types-external.js:397]:
```suggestion
 * @property {(bundlecap: BundleCap, options?: VatUpgradeOptions) => Promise<VatUpgradeResults>} upgrade
 * Restart the vat with the specified bundle and options.
```

File: packages/vats/src/core/basic-behaviors.js
Comment (gibson042) [inline packages/vats/src/core/basic-behaviors.js:296]: I think uses of `VatInfo` in this file (including its definition on line 65) should be replaced with the new `CreateVatResults` type.

**Review by gibson042 — APPROVED** [review]
I agree with @dckc; I don't know how long this would have taken me but I _do_ at least understand it! 🤘

File: packages/SwingSet/src/types-external.js
Comment (warner) [reply packages/SwingSet/src/types-external.js]: good catch, the types were written from the POV of a zoe user, but yes they're definitely not contract-specific

File: packages/SwingSet/src/types-external.js
Comment (warner) [reply packages/SwingSet/src/types-external.js:385]: Yeah, I've had that problem too. In conversation, I've tried to use "admin facet" to refer to the single-vat thing, and "VatAdminService" to refer to the kernel-wide launcher service. But when moving this I realized that `AdminFacet` would be used in places that lacked the "this is for a single vat" context, and might be less confusing if the word "vat" got thrown in there somewhere. Something like "management facet" might help.

I can't think of a perfect name now, and changing it everywhere would get deeper than this PR really needs to go, so I think I'll leave it for now and we can keep thinking of ways to improve it.

**Review by mhofman — DISMISSED** [review]
Thanks so much for taking this on! Is the expectation to fix all occurrences of bundle source as seen in #6826? Because I'm still seeing about 13 cases of them in a loadgen run.

Requesting change until clarification (not an actual review)

**Comment by warner** [comment]
> Thanks so much for taking this on! Is the expectation to fix all occurrences of bundle source as seen in #6826? Because I'm still seeing about 13 cases of them in a loadgen run.

Oh, bummer. I haven't done an exhaustive search: I don't see any other `zoe.install` happening inside `packages/vats`, but I don't know how post-bootstrap proposals happen, so it doesn't surprise me that there's more work to be done.

Once this lands, I'll look at the loadgen slogs and try to find out where the next batch of bundles are coming from.

**Comment by mhofman** [comment]
> it doesn't surprise me that there's more work to be done

@dckc and I noticed:

https://github.com/Agoric/agoric-sdk/blob/3be1a4b6e15a403679e38fac993697f5c6210830/packages/deploy-script-support/src/coreProposalBehavior.js#L133

and 

https://github.com/Agoric/agoric-sdk/blob/71e8578b308a3ecff72957fa91a6806add466d60/packages/inter-protocol/src/proposals/startPSM.js#L307

**Pull request merged**
